src/linux/system-linux.c

```diff
@@ -151,12 +158,156 @@ frida_system_kill (guint pid)
+   * locations are found to be unsuitable, then a future implementation may seek
+   * to validate an ordered list of potential locations.
+   */
+  if (frida_is_directory_noexec (tmp_dir))
+  {
+    return get_application_directory ();
+  }
+  else
+  {
+    return g_strdup (tmp_dir);
+  }
+}
+
+static gboolean
+frida_is_directory_noexec (const gchar * directory)
+{
+  struct stat directory_stat, current_stat;
+  FILE *fp = NULL;
+  struct mntent mntent, *current;
+
+  /*
+   * Since the mounting of the temporary filesystem as `noexec` is not that
+   * common, we assume that this isn't the case unless we can determine
+   * otherwise definitively. (e.g. any errors whilst determining the status
+   * result in us assuming that the directory is executable).
+   */
+  gboolean noexec = FALSE;
+
+  /*
+   * The implementation of this function is derived from `find_mount_point` in
+   * busybox. It finds the mounted filesystem on which a given directory resides
+   * and checks the mount options for `noexec`.
+   */
+  if (stat(directory, &directory_stat) != 0)
+    goto beach;
+
+  if (!S_ISDIR(directory_stat.st_mode))
+    goto beach;
+
+  fp = setmntent ("/proc/mounts", "r");
+  if (fp == NULL)
+    goto beach;
+
+  for (
+    current = getmntent_r(fp, &mntent, mntent_buf, sizeof (mntent_buf));
+    current != NULL;
+    current = getmntent_r(fp, &mntent, mntent_buf, sizeof (mntent_buf)))
+  {
+    /*
+     * In Linux 2.6 rootfs mount always exists, and it makes sense to always
+     * ignore it.
+     */
+    if (g_strcmp0(current->mnt_fsname, "rootfs") == 0)
+      continue;
+
+    /* If the directory name matches the mountpoint */
+    if (g_strcmp0(directory, current->mnt_dir) == 0)
+    {
+      noexec = frida_is_mount_noexec (current);
+      break;
+    }
+
+    /* If the directory name matches mounted the device name */
+    if (g_strcmp0(directory, current->mnt_fsname) == 0)
+    {
+      noexec = frida_is_mount_noexec (current);
+      break;
+    }
+
+    /* Ignore sysfs, proc, none etc */
+    if (current->mnt_fsname[0] == '/')
+    {
+      /* Match the directory and device dev_t */
+      if (stat(current->mnt_fsname, &current_stat) == 0)
+      {
+        if (current_stat.st_rdev == directory_stat.st_dev)
+        {
+          noexec = frida_is_mount_noexec (current);
+          break;
+        }
+      }
+    }
+
+    /* Match the directory and mount point's dev_t */
+    if (stat(current->mnt_dir, &current_stat) == 0)
+    {
+      if (current_stat.st_dev == directory_stat.st_dev)
+      {
+        noexec = frida_is_mount_noexec (current);
+        break;
+      }
+    }
+  }
+
+beach:
+  if (fp != NULL)
+    endmntent (fp);
+
+  return noexec;
+}
+
+static gboolean
+frida_is_mount_noexec (struct mntent * mntent)
+{
+  if (hasmntopt (mntent, "noexec") == NULL)
+    return FALSE;
+  else
+    return TRUE;
+}
+
+static gchar *
+get_application_directory (void)
```

Comment:
This PR has some style deviations. Basically:
- Functions should be prefixed with `frida_` even if internal.
- Missing spaces around `*`.
- Space before `(`.
- Redundant blocks when if/else only has a single statement.